### ISSUE TITLE
[gatsby-plugin-layout] plugin not holding state consistently

### ISSUE THREAD

**Issue by herecydev** [gatsbyjs/gatsby#16243]
### Description

`gatsby-plugin-layout` is unmounting and rerendering when using a programmatic navigate between pages. In my layouts/index.js file I am using a useState hook like `const [time] = useState(Date.now())` and can observe that occasionally the layout component is rerendered causing the time to be altered.

Demonstration:

In cases when the bug occurs and state is lost, the time would progress:
![image](https://user-images.githubusercontent.com/11328618/62200942-51bb6e80-b37e-11e9-9124-1c9d42f380b5.png)

In cases when the bug does not occur and state is maintaned, the time would not progress:
![image](https://user-images.githubusercontent.com/11328618/62201346-2422f500-b37f-11e9-84eb-4004982ae862.png)

This is absolutely not desirable as the intention is use the plugin to keep state, which is a well documented use case. I can see that when this occurs, the page "flashes white", **not** acting like a SPA app. When the bug **doesn't** occur you can see that the layout is not rerendered and the 2nd page component is rendered in a smooth "SPA" fashion.

I am unable to recreate this with `<Link />`, i.e. seems to be only a navigate problem

### Steps to reproduce

Given an absolutely minimal Gatsby app, i.e. [starter-hello-world](https://github.com/gatsbyjs/gatsby-starter-hello-world) and adding `gatsby-plugin-layout` and three files:

layouts/index.js
```js
import React, { useState } from "react"

const Layout = ({ children }) => {
  const [time] = useState(Date.now())
  console.log(time)

  return <div>{children}</div>
}

export default Layout
```

pages/index.js
```js
import React from "react"
import { navigate } from "gatsby"

export default () => <div onClick={() => navigate("/foo")}>Hello World!</div>
```

pages/foo.js
```js
import React from "react"
import { navigate } from "gatsby"

export default () => <div onClick={() => navigate("/")}>Hello Foo!</div>
```

Navigating between index and foo will reproduce the layout state issue, the console logs will demonstrate that the time can change

### Expected result

State would be preserved between pages when storing in a layout component

### Actual result

State is intermittently lost, roughly 25% of the time

### Environment

```
  System:
    OS: Windows 10
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
  Binaries:
    Yarn: 1.17.3 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: 44.17763.1.0
  npmPackages:
    gatsby: ^2.13.45 => 2.13.45 
    gatsby-plugin-layout: ^1.1.2 => 1.1.2 
```

**Comment by alexluong** [comment]
This is super interesting/weird. So I tried recreating your problem in [codesandbox](https://codesandbox.io/s/gatsby-issue-16243-0c94n), and the bug only happens once in the first page change.

However, when I do have another `Link` component on the page, everything works fine, including changing page with `navigate` component.

So, `gatsby-plugin-layout` uses the API `wrapPageElement`, which basically means a parent component for the pages component.

I tried using `wrapRootElement` instead, and add the date into a context that is used by the page components. On page change, the root component is not re-rendered, so although the white flash still happens, the state is untouched. Maybe that would help your use case? Here is the sample [codesandbox](https://codesandbox.io/s/gatsby-issue-16243-emyjh) for that scenario.

If you haven't already, you can learn more about [Browser APIs](https://www.gatsbyjs.org/docs/browser-apis/) and [SSR APIS](https://www.gatsbyjs.org/docs/ssr-apis/).

---

To sum up, I do think there is a bug somewhere in here.

**Comment by herecydev** [comment]
Yer I did mention that I couldn't recreate the bug with a `Link` component, only `navigate`. Thanks for linking the APIs, I have seen them and the [wrapPageElement](https://www.gatsbyjs.org/docs/browser-apis/#wrapPageElement) should be enough but I will try and test with `wrapRootElement` when I get a chance.

Using React's context like you have is much closer to my "real world example" but I tried to simplify as much as possible.

**Comment by alexluong** [comment]
> Yer I did mention that I couldn't recreate the bug with a `Link` component, only `navigate`.

The interesting thing is that if I have a `Link` component on the page and a `button` that will `navigate` on click, everything works fine, including `navigate`. That's what I'm quite confused about. In my first codesandbox, I have that version commented out. You can try it out for yourself.

**Comment by herecydev** [comment]
Doesn't `Link` preload things where as navigate I suppose can't as it doesn't know what will be requested?

**Comment by alexluong** [comment]
Maybe that's the explanation. That sounds very reasonable. To persist data, you would need to wrap it inside `wrapRootElement` as suggested in the documentation.

**Comment by herecydev** [comment]
Where does it suggest that in the documentation? `wrapPageElement` describes:

> This is useful for setting wrapper component around pages that won’t get unmounted on page change

Which is what `gatsby-plugin-layout` uses and says works for "Storing state when navigating pages".

**Comment by alexluong** [comment]
`wrapRootElement`

> This is useful to setup any Providers component that will wrap your application. For setting persistent UI elements around pages use wrapPageElement.

Does this describe your use case?

**Comment by herecydev** [comment]
Yes, but so does `wrapPageElement`

**Comment by herecydev** [comment]
Any other ideas on this one? Would be good to get a firmer understanding of what's going on and why.

**Comment by herecydev** [comment]
not stale

**Comment by metervara** [comment]
I'm also experiencing this on Gatsby 2.13.73.

Following **alexluong** suggestion I have included hidden _Link_ elements for all my routes, which makes navigate work as expected. (The reason I'm not using _Link_ directly is because my clicks originate from sub-nodes inside an svg + I'm executing some code on click besides doing the route change)

> The interesting thing is that if I have a `Link` component on the page and a `button` that will `navigate` on click, everything works fine, including `navigate`. That's what I'm quite confused about. In my first codesandbox, I have that version commented out. You can try it out for yourself.

**Comment by LekoArts** [comment]
Is this still the case with latest versions of Gatsby?
